### ISSUE TITLE
TS Types dont like Object Notation referencing other components

### ISSUE THREAD

**Issue by MWhite-22** [styled-components/styled-components#4336]
## System:
 - OS: macOS 14.5
 - CPU: (10) arm64 Apple M1 Pro
 - Memory: 69.58 MB / 16.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 20.15.0 - ~/.nvm/versions/node/v20.15.0/bin/node
 - npm: 10.7.0 - ~/.nvm/versions/node/v20.15.0/bin/npm
 - pnpm: 9.5.0 - /opt/homebrew/bin/pnpm
## npmPackages:
 - babel-plugin-styled-components: ^2.1.4 => 2.1.4
 - styled-components: ^6.1.11 => 6.1.11

## Reproduction

When using object syntax and referencing other components, TS is complaining about a selector that uses another StyledComponent. For example:
```ts
const Card = styled.div({
  //...styles
});

const CardContainer = styled.div({
  ///...styles,
  [`&:hover > ${Card}`]: {
    //... styles to apply to all child Card components when the CardContainer is hovered
  }
});
```

It appears to be possible, and even encouraged given this code block here: https://github.com/styled-components/styled-components/blob/main/packages/styled-components/src/types.ts#L231-L253

though in IDE there's an issue with the TS types not being assignable to a giant mess of an object
```
Type '{ [x: string]: {}; backgroundColor: "rgba(50, 50, 50, .1)"; }' is not assignable to type 'string | number | RuleSet<any> | StyleFunction<FastOmit<Omit<ScrollAreaScrollbarProps & RefAttributes<HTMLDivElement>, "ref"> & { ...; }, never>> | StyledObject<...> | undefined'.
```

To be clear, it actually works as expected (as long as I also `.toString()` the Styled Component), but the TS types are very unhappy.

**Comment by quantizor** [comment]
This works in v6.3.x — the `${Card}` interpolation in object syntax compiles correctly because styled components intersect with `string` for computed property name compatibility.
